rfcs/012-session-persistence/012-session-persistence.md

```diff
@@ -0,0 +1,303 @@
+   synchronization issues of autosave branches as well as eliminating unnecessary objects polluting LFS.
+2. Inform clients about the state of the repository *inside the PV* and to manipulate it to match what the user wants. 
+3. Present the user with simple choices when the requested session does not match the state inside the PV; allow users
+   to take informed decisions when there is no obvious way forward
+
+### Additional Technical Details
+
+The main difference that we would introduce in Amalthea/Sessions is the following:
+
+1. Sessions exist in one of 3 forms:
+   - **Running**: This is the same as the current sessions
+       - Containers: `oauth2-proxy`, `git-proxy`, `sidecar`, `session`
+       - No changes from currently running sessions
+   - **Hibernate**:
+       - Storage is retained, but the user’s container is gone; the user’s data and code are retained, so we do not need
+         to create autosave branches anymore.
+       - The `sidecar` container is kept to provide information about the repository in the PV; it is used by the
+         clients (UI) to query the state of the user repo (is the repo dirty, what commit it is on, what branch, etc.).
+       - The `oauth2-proxy` container needs to stay for authenticating calls to the sidecar.
+       - The `git-proxy` container needs to stay to inject credentials in calls the sidecar makes.
+       - Users will see sessions in this **Hibernate** form. The main reason for this is that we must surface information
+         about when the **Hibernate** form (and any of its data) will be removed. We should define what actions users will
+         be able to do to these **Hibernate** sessions because they will differ from what is available to **Running** forms.
+   - **Deleted**:
+       - The **Hibernate** version is deleted after some period of not being used. The user should be somehow notified if
+         there is unsaved data here that will be deleted.
+         The UI could mark the sessions about to expire with a warning sign; this might be more effective once we have
+         the dashboard in place because the sessions should be in a primary spot.
+       - This culling interval is different from the culling interval for the **Running** sessions.
+
+2. Shutting down a session (i.e. the user clicking the button in the UI to stop a session) results always in simply
+   going to the **Hibernate** form. In the case that the PV state is fully synced in git, we could consider removing the PV;
+   keeping it around means, however, that we can start the session quicker next time (e.g. the data is already fetched).
+
+3. Culling active sessions turn them into **Hibernate** form. We can consider reducing the culling interval since part of
+   the motivation for keeping it long is to avoid too many autosave branches.
+ 
+4. Additional logic will cull the **Hibernate** forms with a separate (longer) culling interval.
+
```

Comment:
I would add a diagram representing the session lifecycle for clarity:

```mermaid
graph TD
  no-session(No Session)
  starting(Starting)
  running(Running)
  stopping(Stopping)
  hibernating(Hibernating)
  removing(Removing)
  deleted(Deleted)
  error(Error)

  no-session --> |start| starting
  starting --> running
  starting --> |error| error
  running --> |stop| stopping
  running --> |culled| stopping
  stopping --> hibernating
  stopping --> |error| error
  hibernating --> |start| starting
  hibernating --> |remove| removing
  hibernating --> |culled| removing
  removing --> |error| error
  removing --> deleted
```

Reply:
Thanks for the diagram. I've modified it a bit and added it to the RFC.